lib/auth/init.go

```diff
@@ -1020,6 +1028,57 @@ func migrateCertAuthorities(ctx context.Context, asrv *Server) error {
 	return nil
 }
 
+// migrateDBAuthority copies Host CA as Database CA. Before v9.0 database access was using host CA to sign all
+// DB certificates. In order to support existing installations Teleport copies Host CA as Database CA on
+// the first run after update to v9.0+.
+// Function does nothing for databases created with Teleport v9.0+.
+// https://github.com/gravitational/teleport/issues/5029
+func migrateDBAuthority(asrv *Server) error {
+	clusterName, err := asrv.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	dbCaID := types.CertAuthID{Type: types.DatabaseCA, DomainName: clusterName.GetClusterName()}
+	_, err = asrv.GetCertAuthority(dbCaID, false)
+	if err == nil {
+		return nil // no migration needed. DB cert already exists.
+	}
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	// Database CA doesn't exist, check for Host.
+	hostCaID := types.CertAuthID{Type: types.HostCA, DomainName: clusterName.GetClusterName()}
+	hostCA, err := asrv.GetCertAuthority(hostCaID, true)
+	if trace.IsNotFound(err) {
+		// DB CA and Host CA are missing. Looks like the first start. No migration needed.
+		return nil
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Database CA is missing, but Host CA has been found. Database was created with pre v9.
+	// Copy the Host CA as Database CA.
+	log.Infof("Migrating Database CA")
+
+	cav2, ok := hostCA.(*types.CertAuthorityV2)
+	if !ok {
+		return trace.Errorf("Failed to cast Host CA to concrete type.")
+	}
+
+	// Copy the Host CA with a different type.
+	cav2.Spec.Type = types.DatabaseCA
+	cav2.SetSubKind(string(types.DatabaseCA))
+
+	err = asrv.Trust.CreateCertAuthority(cav2)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
We usually recommend scaling auth servers to one before upgrading but not sure how many people actually do that. So there is a possibility of a race if multiple auth servers are attempting the migration concurrently - one will create the CA here and then the other one will get an error. I would handle AlreadyExists error here.

Reply:
Added